packages/core/contracts/financial-templates/common/EmergencyShutdownable.sol

```diff
@@ -0,0 +1,71 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "../../oracle/interfaces/AdministrateeInterface.sol";
+
+/**
+ * @title EmergencyShutdownable contract.
+ * @notice Provides emergency shutdown timestamp and modifiers for the Perpetual contract.
+ */
+
+abstract contract EmergencyShutdownable is AdministrateeInterface {
+    using SafeMath for uint256;
+
+    /****************************************
+     * EMERGENCY SHUTDOWN DATA STRUCTURES *
+     ****************************************/
+
+    // Timestamp used in case of emergency shutdown. 0 if no shutdown has been triggered.
+    uint256 public emergencyShutdownTimestamp;
+
+    /****************************************
+     *                EVENTS                *
+     ****************************************/
+
+    event EmergencyShutdown(address indexed caller, uint256 shutdownTimestamp);
+
+    /****************************************
+     *              MODIFIERS               *
+     ****************************************/
+
+    modifier notEmergencyShutdown() {
+        _notEmergencyShutdown();
+        _;
+    }
+
+    modifier isEmergencyShutdown() {
+        _isEmergencyShutdown();
+        _;
+    }
+
+    /****************************************
+     *          EXTERNAL FUNCTIONS          *
+     ****************************************/
+
+    constructor() public {
+        emergencyShutdownTimestamp = 0;
+    }
+
+    /**
+     * @notice Premature contract settlement under emergency circumstances.
+     */
+    function emergencyShutdown() external virtual override notEmergencyShutdown() {
+        emit EmergencyShutdown(msg.sender, emergencyShutdownTimestamp);
+    }
```

Comment:
Two questions:
1. Is the expectation that this is overridden to include any necessary settlement logic?
2. Should this default implementation set the emergency shutdown timestamp/

Reply:
1. Yes this should be overridden, so maybe its clearer if we don't implement it for now.
2. The reason why it doesn't set the timestamp is because the contract would need to inherit `Testable` in order to call `getCurrentTime()` but this would throw inheritance errors in `FundingRateApplier` which inherits `Testable` from both FeePayer and this contract.


I think the cleanest solution now is to remove the `emergencyShutdown()` method from this contract.

Reply:
Seems reasonable. I don't see a super strong argument either way. Other options:
1. Remove Testable from FeePayer since it will be inherited through here anyway.
2. Declare this method, but don't implement it, which would _require_ derived classes to implement it.

Reply:
1. Then the ExpiringMultiParty won't be `Testable` because the EMP does not use this contract
2. I think this is reasonable but I couldn't think of a good enough reason to do it. It seems like the main point of this contract is to just provide the `emergencyShutdownTimestamp` and modifiers that check its state, but it leave all of the logic surrounding the timestamp to the child.